packages/protocol/contracts/libs/LibTrieProof.sol

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.24;
+
+import {RLPReader} from "../thirdparty/optimism/rlp/RLPReader.sol";
+import {SecureMerkleTrie} from "../thirdparty/optimism/trie/SecureMerkleTrie.sol";
+
+/**
+ * @title LibTrieProof
+ */
+library LibTrieProof {
+    /*********************
+     * Constants         *
+     *********************/
+
+    // The consensus format representing account is RLP encoded in the
+    // following order: nonce, balance, storageHash, codeHash.
+    uint256 private constant ACCOUNT_FIELD_INDEX_STORAGE_HASH = 2;
+
+    /*********************
+     * Public Functions  *
+     *********************/
+
+    /**
+     * Verifies that the value of a slot in the storage of an account is value.
+     *
+     * @param stateRoot The merkle root of state tree.
+     * @param addr The address of contract.
+     * @param slot The slot in the contract.
+     * @param value The value to be verified.
+     * @param mkproof The proof obtained by encoding storage proof.
+     * @return verified The verification result.
+     */
+    function verifyWithAccountProof(
+        bytes32 stateRoot,
+        address addr,
+        bytes32 slot,
+        bytes32 value,
+        bytes calldata mkproof
+    ) public pure returns (bool verified) {
+        (bytes[] memory accountProof, bytes[] memory storageProof) = abi.decode(
+            mkproof,
+            (bytes[], bytes[])
+        );
+
+        bytes memory rlpAccount = SecureMerkleTrie.get(
+            abi.encodePacked(addr),
+            accountProof,
+            stateRoot
+        );
```

Comment:
I would still make is possible to cache the signal root's state root for major costs savings: https://github.com/taikoxyz/taiko-mono/pull/15671#discussion_r1481408404

Reply:
> I would still make is possible to cache the signal root's state root for major costs savings: [#15671 (comment)](https://github.com/taikoxyz/taiko-mono/pull/15671#discussion_r1481408404)

Will do 🖖 , thanks !

Reply:
> I would still make is possible to cache the signal root's state root for major costs savings: [#15671 (comment)](https://github.com/taikoxyz/taiko-mono/pull/15671#discussion_r1481408404)

Brecht, did you mean the SignalService's `storageRoot`? I think it will be a good idea, but we need to make sure the storageRoot is paird with:
- the signalService's chain id
- the signalService's address.

using something like mapping(uint signalServiceChainId => mapping (address signalServiceAddress => mapping (bytes32 storageRoot => bool verified)) verifiedStorageRoot.

Reply:
No, i think it is a `stateRoot` to `storageRoot` mapping. Because we go with full merkle proof, we need storageRoot first then storageRoot and we can cached this 'pair'. And storageRoot is just bytes32 so in case we already having it - relayer no need to submit the signal service root. (account state root).
But yes i guess we can pair the chainId and the signalService address.

Reply:
Lets move caching related conversation to other PRS and try to get this PR merged first.